packages/block-editor/src/hooks/custom-class-name.js

```diff
@@ -55,27 +94,111 @@ export const withInspectorControl = createHigherOrderComponent(
 				'customClassName',
 				true
 			);
+
+			const { updateBlockAttributes } = useDispatch( blockEditorStore );
+
+			const blockStyles = useSelect(
+				( select ) =>
+					select( blocksStore ).getBlockStyles( props.name ),
+				[ props.name, props.attributes.className ]
+			);
+
+			const hasBlockStyles = blockStyles && !! blockStyles.length;
+
+			const activeStyle = hasBlockStyles
+				? getActiveStyle(
+						blockStyles,
+						props.attributes.className || ''
+				  )
+				: null;
+
+			const onSelectStyleClassName = ( style ) => {
+				const styleClassName = replaceActiveStyle(
+					props.attributes.className,
+					activeStyle,
+					style
+				);
+				updateBlockAttributes( props.clientId, {
+					className: styleClassName,
+				} );
+			};
+
+			const additionalClassNameContainerClasses = classnames(
+				'additional-class-name-control__container',
+				{
+					'has-block-styles': hasBlockStyles,
+				}
+			);
+
 			if ( hasCustomClassName && props.isSelected ) {
 				return (
 					<>
 						<BlockEdit { ...props } />
 						<InspectorControls __experimentalGroup="advanced">
-							<TextControl
-								autoComplete="off"
-								label={ __( 'Additional CSS class(es)' ) }
-								value={ props.attributes.className || '' }
-								onChange={ ( nextValue ) => {
-									props.setAttributes( {
-										className:
-											nextValue !== ''
-												? nextValue
-												: undefined,
-									} );
-								} }
-								help={ __(
-									'Separate multiple classes with spaces.'
-								) }
-							/>
+							<div
+								className={
+									additionalClassNameContainerClasses
+								}
+							>
+								<TextControl
+									className="additional-class-name-control__text-control"
+									autoComplete="off"
+									label={ __( 'Additional CSS class(es)' ) }
+									value={ props.attributes.className || '' }
+									onChange={ ( nextValue ) => {
+										props.setAttributes( {
+											className:
+												nextValue !== ''
+													? nextValue
+													: undefined,
+										} );
+									} }
+									help={ __(
+										'Separate multiple classes with spaces.'
+									) }
+								>
+									{ hasBlockStyles && (
+										<DropdownMenu
```

Comment:
Would it be better to split this out into its own component? That might help reduce the depth of component nesting we are ending up with here.

Reply:
Thanks for the 👀 

You're making sense. I like it.

Reply:
fd15e64f5cf4f2a066c9da97cb34073879cd98f6 splits out the custom inspector controls into a component 👍